### PR TITLE
Update built-in fields to new validate hook syntax

### LINKED ISSUE #9165: Unable to use new `validate.{create|update|delete}` field-level hooks with most built-in fields
I looked around for a similar issue/discussion but didn't see one, apologies if this is a duplicate.

At the moment, it appears to be impossible to use the new `validate.{create|update|delete}` field-level hooks with most of the built-in Keystone field types.

This appears to be due to [this list check](https://github.com/keystonejs/keystone/blob/main/packages/core/src/lib/core/initialise-lists.ts#L314-L316) which asserts the mutual exclusivity of the new validate syntax and the old validate syntax.

Most of the built-in fields define a `validateInput` check that's automatically added if `db.isNullable` is set to false or if you specify validation options ([example](https://github.com/keystonejs/keystone/blob/main/packages/core/src/fields/types/text/index.ts#L117)). Attempting to define a custom validate.create/validate.update check results in the following error:

```
TypeError: "hooks.validate" conflicts with "hooks.validateInput" for the "name" field
    at parseFieldHooks (myapp/node_modules/@keystone-6/core/dist/createSystem-b6c61fe6.cjs.dev.js:3793:50)
    at getListsWithInitialisedFields (myapp/node_modules/@keystone-6/core/dist/createSystem-b6c61fe6.cjs.dev.js:3871:16)
    at initialiseLists (myapp/node_modules/@keystone-6/core/dist/createSystem-b6c61fe6.cjs.dev.js:4305:25)
    at Object.createSystem (myapp/node_modules/@keystone-6/core/dist/createSystem-b6c61fe6.cjs.dev.js:4548:17)
```

Given that the validateInput/validateDelete syntax has been marked as deprecated (at least code comments seem to indicate that), I expected the built-in fields to allow usage of the new syntax (maybe it would make sense to have them automatically switch based on the presence of the validate object?)

Happy to submit a PR for this, but figured I'd also file an issue.

Keystone Version: latest (6.1.0)
Node Version: v20

### PR DESCRIPTION
**Pull request by acburdine**
closes #9165
- this updates the built-in fields using validateInput to use the new unified validation syntax, and handles merging with the old syntax to ensure user-land old syntax still works

### CODE REVIEW

**Comment by dcousens** [comment]
Amazing @acburdine, this is tedious and really appreciated work! :green_heart: 
I'll review, merge + release this as soon as I can.

**Comment by gautamsi** [comment]
I am working on #9204 which removes the deprecated hooks and cleanups code for Fields as well as consuming code in a breaking changes. This PR will not be needed if that approach works for @dcousens

**Comment by dcousens** [comment]
@gautamsi https://github.com/keystonejs/keystone/pull/9204 will be a breaking change, which is OK, we wanted that eventually, but, we should let users gracefully upgrade by releasing #9166 (this) first

**Comment by gautamsi** [comment]
we can do that, i guess the other one would be part of v-next branch and might take longer, I agree with you to get this going.

File: packages/core/src/fields/resolve-hooks.ts [hunk 1]
```diff
@@ -11,68 +23,47 @@ function splitValidateHooks <ListTypeInfo extends BaseListTypeInfo> (
...
   if (builtin === undefined) return hooks
 
-  const result: FieldHooks<ListTypeInfo> = {
-    resolveInput: hooks?.resolveInput,
-  }
-
-  if (hooks.beforeOperation || builtin.beforeOperation) {
-    result.beforeOperation = async (args) => {
-      await hooks.beforeOperation?.(args)
-      await builtin.beforeOperation?.(args)
```
Comment (dcousens) [inline packages/core/src/fields/resolve-hooks.ts]: @acburdine quick question, did you have any intent behind `hooks.beforeOperation` being first?

File: tests/api-tests/fields/types/fixtures/decimal/test-fixtures.ts [hunk 2]
```diff
@@ -55,7 +55,7 @@ export const crudTests = (keystoneTestWrapper: any) => {
       expect(result.errors).toHaveLength(1)
       expect(result.errors![0].message).toMatchInlineSnapshot(`
         "You provided invalid data for this operation.
-          - Test.price: Price must be greater than or equal to -300"
+          - Test.price: value must be greater than or equal to -300"
```
Comment (dcousens) [inline tests/api-tests/fields/types/fixtures/decimal/test-fixtures.ts:58]: Errors are already prefixed with `{listKey}.{fieldKey}`, this humanized part was redundant and became difficult to understand for longer names

File: packages/core/src/fields/types/bigInt/index.ts [hunk 3]
```diff
@@ -30,67 +29,87 @@ export type BigIntFieldConfig<ListTypeInfo extends BaseListTypeInfo> =
...
     }
-    if (validation.min > validation.max) {
-      throw new Error(`The bigInt field at ${meta.listKey}.${meta.fieldKey} specifies a validation.max that is less than the validation.min, and therefore has no valid options`)
+    if (
+      min !== undefined &&
+      max !== undefined &&
+      min > max
+    ) {
+      throw new Error(`${meta.listKey}.${meta.fieldKey} specifies a validation.max that is less than the validation.min, and therefore has no valid options`)
     }
```
Comment (dcousens) [inline packages/core/src/fields/types/bigInt/index.ts:87]: I leveraged this opportunity to unify each of the numerical fields to share nearly the same code for `min` and `max` validation.

**Review by dcousens — APPROVED** [review]

File: packages/core/src/fields/resolve-hooks.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    ...hooks,
+    beforeOperation: merge(builtin.beforeOperation, hooks.beforeOperation),
+    afterOperation: merge(builtin.afterOperation, hooks.afterOperation),
+    validate: (builtinValidate || hooksValidate) ? {
+      create: merge(builtinValidate?.create, hooksValidate?.create),
+      update: merge(builtinValidate?.update, hooksValidate?.update),
+      delete: merge(builtinValidate?.delete, hooksValidate?.delete)
+    } : undefined,
+  } satisfies FieldHooks<ListTypeInfo>
+}
```
Comment (dcousens) [inline packages/core/src/fields/resolve-hooks.ts:69]: @acburdine I tidied this up a little, but fundamentally it's the same

File: packages/core/src/fields/resolve-hooks.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  hooks?: FieldHooks<ListTypeInfo>,
+) {
+  if (hooks === undefined) return builtin
+  if (builtin === undefined) return hooks
+
+  const builtinValidate = resolveValidateHooks(builtin)
+  const hooksValidate = resolveValidateHooks(hooks)
+  return {
+    ...hooks,
+    beforeOperation: merge(builtin.beforeOperation, hooks.beforeOperation),
```
Comment (dcousens) [inline packages/core/src/fields/resolve-hooks.ts:61]: @acburdine I did notice you had `hooks.beforeOperation` before the built-ins, was that intentional?

Comment (dcousens) [reply packages/core/src/fields/resolve-hooks.ts]: Resolved, this was intentional, and I have added a comment as such

File: tests/api-tests/fields/types/fixtures/text/test-fixtures.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ export const name = 'Text'
 export const typeFunction = text
 export const exampleValue = () => 'foo'
 export const exampleValue2 = () => 'bar'
+export const nonNullableDefault = true
```
Comment (dcousens) [inline tests/api-tests/fields/types/fixtures/text/test-fixtures.ts:7]: By adding this for `multiselect`, and `Text` supporting `null` in https://github.com/keystonejs/keystone/pull/9057, `text` is now _less_ special in our tests

File: tests/api-tests/fields/required.test.ts [hunk 7]
```diff
@@ -53,6 +52,11 @@ for (const modulePath of testModules) {
               fields: {
                 name: text(),
                 testField: mod.typeFunction({
+                  ...(mod.nonNullableDefault ? {
+                    db: {
+                      isNullable: true
```
Comment (dcousens) [inline tests/api-tests/fields/required.test.ts:57]: These tests are written around the premise that `db.isNullable` is `true` by default

**Comment by dcousens** [comment]
Thanks for starting this work @acburdine, this has been a heavy lift required for a while, and I don't think I could have jumped in to bring the fields into this kind of alignment around defaults, `isRequired` and `db.isNulalble` as this pull request has now without your help!

Thank you :yellow_heart:

**Pull request merged**